src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.cs

```diff
@@ -168,49 +207,84 @@ internal partial class AnalyzerManager
-                Action<Exception, DiagnosticAnalyzer, Diagnostic> onAnalyzerException = analyzerExecutor.OnAnalyzerException;
-                if (onAnalyzerException != null)
+                // Subscribe for exceptions from lazily evaluated localizable strings in the descriptors.
+                foreach (var descriptor in supportedDiagnostics)
                 {
-                    handler = new EventHandler<Exception>((sender, ex) =>
-                    {
-                        var diagnostic = AnalyzerExecutor.CreateAnalyzerExceptionDiagnostic(analyzer, ex);
-                        onAnalyzerException(ex, analyzer, diagnostic);
-                    });
+                    descriptor.Title.OnException += handler;
+                    descriptor.MessageFormat.OnException += handler;
+                    descriptor.Description.OnException += handler;
+                }
+            }
 
-                    // Subscribe for exceptions from lazily evaluated localizable strings in the descriptors.
-                    foreach (var descriptor in supportedDiagnostics)
-                    {
-                        descriptor.Title.OnException += handler;
-                        descriptor.MessageFormat.OnException += handler;
-                        descriptor.Description.OnException += handler;
-                    }
+            return Tuple.Create(supportedDiagnostics, handler);
+        }
+
+        /// <summary>
+        /// Return <see cref="DiagnosticAnalyzer.SupportedDiagnostics"/> of given <paramref name="analyzer"/>.
+        /// </summary>
+        public ImmutableArray<DiagnosticDescriptor> GetSupportedDiagnosticDescriptors(
+            DiagnosticAnalyzer analyzer,
+            AnalyzerExecutor analyzerExecutor)
+        {
+            // Check if the value has already been computed and stored.
+            Tuple<ImmutableArray<DiagnosticDescriptor>, EventHandler<Exception>> value;
+            lock (_descriptorMap)
+            {
+                if (_descriptorMap.TryGetValue(analyzer, out value))
+                {
+                    return value.Item1;
                 }
+            }
 
-                return Tuple.Create(supportedDiagnostics, handler);
-            });
+            // Otherwise, compute the value.
+            // We do so outside the lock statement as we are calling into user code, which may be a long running operation.
+            value = ComputeDescriptorsAndHandler(analyzer, analyzerExecutor);
 
-            return descriptors.Item1;
+            lock (_descriptorMap)
+            {
+                // Check if another thread already stored the computed value.
+                Tuple<ImmutableArray<DiagnosticDescriptor>, EventHandler<Exception>> storedValue;
+                if (_descriptorMap.TryGetValue(analyzer, out storedValue))
+                {
+                    // If so, we return the stored value.
+                    value = storedValue;
```

Comment:
Unfortunately, it won't be easy to test it as it will require 2 analysis threads to simultaneously request descriptors for the same analyzer, such that both of them go past the first lock before any of them enter the second lock.